pkg/api/track.go

```diff
@@ -52,68 +53,202 @@ func (vwo *VWOInstance) Track(campaignKey, userID, goalIdentifier string, option
 	if err != nil {
-		message := fmt.Sprintf(constants.ErrorMessageTrackAPIGoalNotFound, vwoInstance.API, goalIdentifier, campaignKey, userID, err.Error())
-		utils.LogMessage(vwo.Logger, constants.Error, track, message)
+		message := fmt.Sprintf(constants.ErrorMessageTrackAPIGoalNotFound, vwoInstance.API, goalIdentifier, campaign.Key, userID, err.Error())
+		utils.LogMessage(vwoInstance.Logger, constants.Error, track, message)
 		return false
 	}
 
-	if goal.Type == constants.GoalTypeRevenue && options.RevenueValue == nil  {
-		message := fmt.Sprintf(constants.ErrorMessageTrackAPIRevenueNotPassedForRevenueValue, vwoInstance.API, goalIdentifier, campaignKey, userID)
-		utils.LogMessage(vwo.Logger, constants.Error, track, message)
+	if goal.Type != goalTypeToTrack && goalTypeToTrack != constants.GoalTypeAll {
+		message := fmt.Sprintf(constants.ErrorMessageInvalidGoalType, vwoInstance.API, goalTypeToTrack, goal.Type)
+		utils.LogMessage(vwoInstance.Logger, constants.Error, track, message)
 		return false
 	}
 
-	variation, err := core.GetVariation(vwoInstance, userID, campaign, options)
-	if err != nil {
-		message := fmt.Sprintf(constants.InfoMessageInvalidVariationKey, vwoInstance.API, userID, campaignKey, err.Error())
-		utils.LogMessage(vwo.Logger, constants.Info, track, message)
+	if goal.Type == constants.GoalTypeRevenue && options.RevenueValue == nil {
+		message := fmt.Sprintf(constants.ErrorMessageTrackAPIRevenueNotPassedForRevenueValue, vwoInstance.API, goalIdentifier, campaign.Key, userID)
+		utils.LogMessage(vwoInstance.Logger, constants.Error, track, message)
 		return false
 	}
 
-	impression := utils.CreateImpressionTrackingGoal(vwoInstance, variation.ID, userID, goal.Type, campaign.ID, goal.ID, options.RevenueValue)
-	go event.DispatchTrackingGoal(vwoInstance, goal.Type, impression)
-
-	message := fmt.Sprintf(constants.InfoMessageMainKeysForImpression, vwoInstance.API, vwoInstance.SettingsFile.AccountID, vwoInstance.UserID, campaign.ID, variation.ID)
-	utils.LogMessage(vwo.Logger, constants.Info, activate, message)
+	variation, storedGoalIdentifier, err := core.GetVariation(vwoInstance, userID, campaign, goalIdentifier, options)
+	if err != nil {
+		message := fmt.Sprintf(constants.InfoMessageInvalidVariationKey, vwoInstance.API, userID, campaign.Key, err.Error())
+		utils.LogMessage(vwoInstance.Logger, constants.Info, track, message)
+		return false
+	}
+	
+	if variation.Name != "" {
+		if storedGoalIdentifier != "" {
+			identifiers := strings.Split(storedGoalIdentifier, constants.GoalIdentifierSeperator)
+			flag := false
+
+			for _, identifier := range identifiers {
+				if identifier == goalIdentifier {
+					flag = true
+				}
+			}
+
+			fmt.Println(flag)
+			if flag == false {
+				storedGoalIdentifier += constants.GoalIdentifierSeperator + goalIdentifier
+
+				if storage, ok := vwoInstance.UserStorage.(interface{ Set(a, b, c, d string) }); ok {
+					storage.Set(userID, campaign.Key, variation.Name, storedGoalIdentifier)
```

Comment:
you are only setting storedGoalIdentifier when its previously set,

no code for setting it the first time if userStorage exists but goalIdentifier is not set.

Reply:
you should maybe do it at line 230,

Check if userStorage exists, set goalIdentifier as `storedGoalIdentifier`

Reply:
ignore, see it on line 234